crates/driver/src/infra/token_info/mod.rs

```diff
@@ -0,0 +1 @@
+pub use crate::boundary::token_info::Fetcher;
```

Comment:
To be consistent with the rest of the design, the `infra` code should operate on `domain` types and not `boundary` types (i.e. `TokenInfo`).

As such, the `infra::Fetcher` should probably return a `domain::competition::Tokens` and not a `boundary::token_info::TokenInfo`. Also, I think the `boundary` model should serve as an adapter between legacy code, and the rest of the driver code.

Personally, I would port the token fetching to the new driver architecture instead of using a `boundary` module since it is small enough (and then we can delete the "token info fetching" from `shared`, `orderbook` and `autopilot` once we fully co-locate).

Reply:
I ended up implementing the `Fetcher` in `infra` directly so we don't use more `shared` code.
Things to note:
* did not use a trait
* used `RwLock` instead of `tokio::Mutex` to not block while making requests to the node
* dropped the whole `BatchCall` setup since those calls should be batched under the hood by your fancy `Web3` extension (although I believe that this still needs to be used in the driver)

Reply:
> did not use a trait

Good - I don't think we should have mocking. That being said - this should be reflected in an E2E test IMO.

> used `RwLock` instead of `tokio::Mutex` to not block while making requests to the node

Sounds reasonable. You may double-fetch some tokens this way, but I think that is worth not delaying the fetch too long (which makes more sense in the context of using a `driver` for quoting as well).

> dropped the whole `BatchCall` setup

Nice.

> exposed `web3` type from Blockchain

Not a fan of this, left a comment on the approach I would use (spoiler - I think it should use `eth.contract_at::<ERC20>(token_address)`).